File: src/main/scala/Day1/Day1.scala
package Day1

import scala.io.Source

object Day1 {
    def main(args: Array[String]): Unit = {
        val source = Source.fromFile("src/main/scala/Day1/input")
        val xs = source.getLines().toList.map(_.toInt)
        source.close()

        part1(xs)
        part2(xs)
    }

    private def part1(lines: List[Int]): Unit = {
        val func = lines.sliding(2).count {
            case List(a, b) => a < b
            case _ => false
        }
        println(
            func + " (1791)"
        )
    }

    private def part2(lines: List[Int]): Unit = {
        val func: Int = lines.sliding(3).map(_.sum).toList.sliding(2).count {
            case List(a, b) => a < b
            case _ => false
        }
        println(func + " (1822)")
    }

}

File: src/main/scala/Day1/Day1.hs
main = do
   ls <- fmap lines (readFile "Day1/input")
   let intList = map (read::String->Int) ls
   let pairList = makePairs intList

   print $ part1 pairList
   print $ part2 intList

part1 :: (Num a, Ord a) => [(a,a)] -> Int
part1 xs = length (filter (\(x,y) -> x<y) xs)

part2 :: (Num a, Ord a) => [a] -> Int
part2 xs = let tripleSums =  map (sum) $ windows3 xs xs xs
        in part1 $ makePairs tripleSums


{-Erstellt Paare nach dem Schema [1,2,3,4] => [(1,2),(2,3),(3,4)] -}
makePairs :: [a] -> [(a,a)]
makePairs xs = zip xs (tail xs)
    
{-Erstellt eine Liste vom Triples. Analog zu bekanntem `sliding(3)`-}
windows3 :: [a] -> [a] -> [a] -> [[a]]
windows3 (xs) (ys) (zs) 
    | length xs > 0 && length ys >1 && length zs >2 = [xs!!0, ys!!1, zs!!2] : windows3 (tail xs) (tail ys) (tail zs)
    | otherwise                 = []                            

